File: Leetcode/Trees/437. Path Sum III.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int pathSum(TreeNode root, int targetSum) {
        if (root == null) return 0;
        else if (root.left == null && root.right == null && root.val == targetSum) return 1;
        else return pathSum(root.left, targetSum - root.val) + pathSum(root.right, targetSum - root.val) + pathSum(root.left, targetSum) + pathSum(root.right, targetSum);
    }
}

File: Leetcode/Trees/437. Path Sum III.py
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        def helper(root, targetSum):
            if root:
                count = 0
                if targetSum == root.val:
                    count += 1
                return count + helper(root.left, targetSum - root.val) + helper(root.right, targetSum - root.val)
            return 0
        L, R = 0, 0
        if root:
            if root.left:
                L = self.pathSum(root.left, targetSum)
            if root.right:
                R = self.pathSum(root.right, targetSum)
        return helper(root, targetSum) + L + R
